src/garbage_collector.rs

```diff
@@ -56,22 +63,50 @@ impl GarbageCollector {
                 db_stats,
             };
             tokio_handle.block_on(orchestrator.run());
+
+            // !important: make sure that this is always the last thing that this
+            // thread does, otherwise we risk notifying waiters and leaving this
+            // thread around after shutdown
+            if shutdown_tx.send(true).is_err() {
+                error!("Could not send shutdown signal to threads blocked on await_shutdown");
+            }
         });
         Self {
-            main_thread: Some(main_thread),
-            main_tx: external_tx,
+            main_tx: Arc::new(external_tx),
+            shutdown_rx,
+        }
+    }
+
+    /// Waits for the main garbage collection thread to complete. This does
+    /// not cause the thread to shut down, use [trigger_shutdown] or [close]
+    /// instead to signal the thread to terminate
+    pub(crate) async fn await_shutdown(mut self) {
+        while !*self.shutdown_rx.borrow() {
+            self.shutdown_rx
+                .changed()
+                .await
+                .expect("Shutdown rx disconnected.");
         }
     }
 
-    /// Close the garbage collector
-    pub(crate) async fn close(mut self) {
-        if let Some(main_thread) = self.main_thread.take() {
-            self.main_tx.send(Shutdown).expect("main tx disconnected");
-            main_thread
-                .join()
-                .expect("failed to stop main compactor thread");
+    /// Triggers the main garbage collection thread to terminate
+    fn trigger_shutdown(&self) {
```

Comment:
Curious why this can't be a `pub(crate)` so we don't need to expose the `main_tx`. I'm guessing it's some Rust annoyance I'm missng?

Reply:
It always is some Rust ownership thing 😆 I had two choices:

1. expose `main_tx` so that we could clone only that when we use it from within the thread that waits for the interrupt:
```rust
    let gc_tx = collector.main_tx.clone();
    tokio::spawn(async move {
        tokio::signal::ctrl_c()
            .await
...
        let _ignored_error = gc_tx.send(Shutdown);
    });
```

2. make the entire `GarbageCollector` implement the `Clone` trait, and be safe to use across threads and pass that in instead:

```rust
    let gc_clone = collector.clone();
    tokio::spawn(async move {
        tokio::signal::ctrl_c()
            .await
...
        let _ignored_error = gc.trigger_shutdown();
    });
```

I felt like (1) was a little better, but I don't feel too strongly. Perhaps the best way is a middle ground of wrapping the `main_tx` in an individually cloneable object and exposing that through some trait (I had thought about doing this, but that felt like overkill)

Reply:
Honestly, though, I'm far from having a well developed intuition on what's better long-term in Rustland. More than happy to defer to someone that has developed a sense for the implications of this... All I can hope for is that I can explain my decision 😆

Reply:
WDYT about moving the spawn logic into g_c.rs?

```
diff --git a/src/admin.rs b/src/admin.rs
index 9187c75..febbed4 100644
--- a/src/admin.rs
+++ b/src/admin.rs
@@ -2,7 +2,6 @@ use crate::checkpoint::Checkpoint;
 use crate::config::GarbageCollectorOptions;
 use crate::error::SlateDBError;
 use crate::garbage_collector::GarbageCollector;
-use crate::garbage_collector::GarbageCollectorMessage::Shutdown;
 use crate::manifest_store::ManifestStore;
 use crate::metrics::DbStats;
 use crate::sst::SsTableFormat;
@@ -16,7 +15,6 @@ use std::error::Error;
 use std::ops::RangeBounds;
 use std::sync::Arc;
 use tokio::runtime::Handle;
-use tracing::debug;
 
 /// read-only access to the latest manifest file
 pub async fn read_manifest(
@@ -109,16 +107,7 @@ pub async fn run_gc_instance(
     )
     .await;
 
-    let gc_tx = collector.main_tx.clone();
-    tokio::spawn(async move {
-        tokio::signal::ctrl_c()
-            .await
-            .expect("Failed to install CTRL+C signal handler");
-        debug!("Intercepted SIGINT ... shutting down garbage collector");
-        // if we cant send a shutdown message it's probably because it's already closed
-        let _ignored_error = gc_tx.send(Shutdown);
-    });
-
+    collector.spawn_shutdown_signal_handler();
     collector.await_shutdown().await;
     Ok(())
 }
diff --git a/src/garbage_collector.rs b/src/garbage_collector.rs
index f85654a..602c248 100644
--- a/src/garbage_collector.rs
+++ b/src/garbage_collector.rs
@@ -5,6 +5,7 @@ use std::sync::Arc;
 use std::time::{Duration, Instant};
 use std::{fmt, thread};
 use tokio::runtime::Handle;
+use tokio::signal;
 use tokio::sync::watch;
 use tracing::{debug, error, info};
 
@@ -26,7 +27,7 @@ pub(crate) enum GarbageCollectorMessage {
 
 #[derive(Clone)]
 pub(crate) struct GarbageCollector {
-    pub(crate) main_tx: Arc<crossbeam_channel::Sender<GarbageCollectorMessage>>,
+    main_tx: Arc<crossbeam_channel::Sender<GarbageCollectorMessage>>,
     shutdown_rx: watch::Receiver<bool>,
 }
 
@@ -92,10 +93,23 @@ impl GarbageCollector {
     /// Triggers the main garbage collection thread to terminate
     fn trigger_shutdown(&self) {
         if self.main_tx.send(Shutdown).is_err() {
-            error!("Could not send shutdown signal to threads blocked on await_shutdown");
+            error!("Could not send shutdown signal to garbage collector thread");
         }
     }
 
+    /// Spawns a task that listens for Ctrl+C and triggers shutdown
+    pub(crate) fn spawn_shutdown_signal_handler(&self) {
+        let main_tx = self.main_tx.clone();
+        tokio::spawn(async move {
+            signal::ctrl_c()
+                .await
+                .expect("Failed to install CTRL+C signal handler");
+            debug!("Intercepted SIGINT ... shutting down garbage collector");
+            // if we cant send a shutdown message it's probably because it's already closed
+            let _ignored_error = main_tx.send(Shutdown);
+        });
+    }
+
     /// Close the garbage collector and await clean termination
     pub(crate) async fn close(self) {
         self.trigger_shutdown();
```

I asked Windsurfer about this, and that's what it came up with. Seems reasonable.

Reply:
yeah, I actually quite like that! Maybe I ought to try Windsurfer, I asked ChatGPT quite a bit about this and didn't get to anything I liked. Probably lacking all the context necessary. I'm impressed.

Reply:
Yea. I am kinda shocked. Here's how it looks:

![Screenshot 2024-11-25 at 3 31 53 PM](https://github.com/user-attachments/assets/7e6339ff-96c1-480d-977b-8870709e622b)

(cropped at bottom due to screen size)

Reply:
https://github.com/slatedb/slatedb/pull/358 - seems so obvious in retrospect

Reply:
🔥 merged!